Modify repo.cmake to support musl libc.a and LLVM libc++.a.

MaggieYingYi opened the pull request:
After porting the musl-libc to the program repository, we want to use the musl-libc to build the project. The repo.cmake file is modified to support of building the project using musl libc library and the LLVM libc++ library.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+	set (libcxxabi_include "${libcxxabi_include}")
+endif ()
+
+# The user may specify the exception and rtti by giving
+# '-Dexception_rtti:BOOL=Yes' on the command line
```
Is this something that we want to allow (given that we don't yet have exception support in the tools)?

MaggieYingYi replied:
Good point, removed the variable and modified in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/6c4f10ca7cb881e1330eb14bea461620683715d0.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+	set (libcxxabi_include "${libcxxabi_include}")
+endif ()
+
+# The user may specify the exception and rtti by giving
+# '-Dexception_rtti:BOOL=Yes' on the command line
```
Is this something that we want to allow (given that we don't yet have exception support in the tools)?

MaggieYingYi replied:
Fixed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/6c4f10ca7cb881e1330eb14bea461620683715d0.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+	set (exceptions_rtti "-fno-exceptions -fno-rtti")
+endif ()
+
+# The user may specify the musl_install by giving
```
It would be more useful to describe what this variable means rather than just that I can set it. Presumably it refers to the directory in which musl-c is installed.

MaggieYingYi replied:
Fixed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/6c4f10ca7cb881e1330eb14bea461620683715d0.

paulhuggett replied:
> musl_install: the path of the directory in which musl libc is installed\

This should be in the future tense, I think: “the path of the directory in which musl libc will be installed". We’re describing the path which will be used by a later install step, aren’t we? It’s not a path that we expect to already exist.

MaggieYingYi replied:
This file is repo.cmake file. We except both the `musl_install` and `llvm_install` already exist/installed and then we can use musl libc and LLVM libraries to build the user's project. Therefore, the `musl_install` and `llvm_install` should not be future tense.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+	set (musl_install "/home/prepo/musl")
+endif ()
+
+# The user may specify the llvm_install by giving
```
As above. The comment should really describe to what the value of the variable refers.

MaggieYingYi replied:
Fixed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/6c4f10ca7cb881e1330eb14bea461620683715d0.

paulhuggett replied:
(As above.)

MaggieYingYi replied:
The same reason is given as above.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+# The user may use the MUSL libc.a by giving
+# '-Dmusl:BOOL=Yes' on the command line
+if (musl)
+	SET (MUSL_COMPILE_FLAGS "-nostdinc -nostdinc++ --sysroot ${musl_install} -isystem ${musl_install}/include ${libcxxabi_include}")
```
All of the other variables are lower-case. Is there a reason why this one is different?

MaggieYingYi replied:
Fixed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/6c4f10ca7cb881e1330eb14bea461620683715d0.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+# '-Dmusl:BOOL=Yes' on the command line
+if (musl)
+	SET (MUSL_COMPILE_FLAGS "-nostdinc -nostdinc++ --sysroot ${musl_install} -isystem ${musl_install}/include ${libcxxabi_include}")
+	SET (CMAKE_EXE_LINKER_FLAGS "-nostdlib -nodefaultlibs -static --sysroot ${musl_install} -L ${musl_install}/lib -L ${llvm_install}/lib ${libcxxabi_lib} -lc" CACHE STRING "toolchain_exelinkflags" FORCE)
```
Is there a deliberate choice in setting this variable rather than the existing  CMAKE_XX_LINK_EXECUTABLE variables below?

MaggieYingYi replied:
`CMAKE_EXE_LINKER_FLAGS` is a Linker flag to be used to create executables. It is used by the `CMAKE_XX_LINK_EXECUTABLE` variables below. `CMAKE_EXE_LINKER_FLAGS` contains the linker options which are used by both `CMAKE_C_LINK_EXECUTABLE` and `CMAKE_CXX_LINK_EXECUTABLE`.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
+	SET (CMAKE_EXE_LINKER_FLAGS "-nostdlib -nodefaultlibs -static --sysroot ${musl_install} -L ${musl_install}/lib -L ${llvm_install}/lib ${libcxxabi_lib} -lc" CACHE STRING "toolchain_exelinkflags" FORCE)
+endif()
+
+SET (CMAKE_C_FLAGS "${MUSL_COMPILE_FLAGS} ${exceptions_rtti}" CACHE STRING "toolchain_cflags")
+SET (CMAKE_CXX_FLAGS "${MUSL_COMPILE_FLAGS} ${exceptions_rtti} ${libcxx}" CACHE STRING "toolchain_cxxflags")
```
I'm not sure about blasting the values of CMAKE_XX_FLAGS. These are set by the user if they want to override the default compilation flags and this will (I think) silently override the user's choice.

MaggieYingYi replied:
I tried to set CMAKE_XX_FLAGS from both command line and repo.cmake file. For example:
```
        cmake -D musl=Yes -D libcxx=Yes                                       \
                -D CMAKE_BUILD_TYPE=Release                                   \
                -D CMAKE_CXX_FLAGS="-frtti -fcxx-exceptions"                  \
                -D LLVM_PATH=../                                              \
                -D LIBUNWIND_ENABLE_SHARED=No                                 \
                -D LIBUNWIND_ENABLE_STATIC=Yes                                \
                -D CMAKE_TOOLCHAIN_FILE=$(TOOLCHAIN_FILE)                     \
                -D CMAKE_INSTALL_PREFIX=$(LLVM)                               \
                ../libunwind &&                                               \
```

I found the `CMAKE_CXX_FLAGS` is set to the "-frtti -fcxx-exceptions" instead of the value in the repo.cmake file. Therefore, the CMAKE_XX_FLAGS value in the repo.cmake file will not override the user's choice in the command line.

paulhuggett commented on llvm/utils/repo/repo.cmake:
```diff
@@ -24,10 +24,57 @@ else ()
-SET (CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O0 -fno-exceptions -fno-rtti" CACHE STRING "Default CXX Flags Debug")
-SET (CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O3 -fno-exceptions -fno-rtti" CACHE STRING "Default C Flags Release")
-SET (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -fno-exceptions -fno-rtti" CACHE STRING "Default CXX Flags Release")
+# The user may specify the libcxxabi_include by giving
```
As elsewhere, it would be more useful to describe the function of the variable here.

MaggieYingYi replied:
Fixed in the commit [6c4f10c](https://github.com/SNSystems/llvm-project-prepo/commit/6c4f10ca7cb881e1330eb14bea461620683715d0).

paulhuggett reviewed (commented):
I think that this is a good approach to the problem of supporting the new C/C++ stdlibs. Can I confirm that the instructions for building LLVM and pstore with the docker image are still good with this revised toolchain file?

MaggieYingYi commented:
I added a new Makefile to build LLVM libraries using the musl-prepo libc library in commit  https://github.com/SNSystems/llvm-project-prepo/commit/6c4f10ca7cb881e1330eb14bea461620683715d0 and install it in the docker in commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/2a5b14bcf80f6bba91381b6b09507c09f571645f.

paulhuggett commented:
> I added a new Makefile to build LLVM libraries 

When you say ”LLVM libraries”, which libraries do you mean? I’m guessing that this makefile doesn’t build the compiler.
  
> using the musl-prepo libc library in commit [6c4f10c](https://github.com/SNSystems/llvm-project-prepo/commit/6c4f10ca7cb881e1330eb14bea461620683715d0) and install it in the docker in commit [2a5b14b](https://github.com/SNSystems/llvm-project-prepo/commit/2a5b14bcf80f6bba91381b6b09507c09f571645f).

Why is it being added to the docker image? (We don't have the code in there, or the equivalent for the other libraries.)

paulhuggett commented on llvm/utils/repo/Makefile:
```diff
@@ -0,0 +1,121 @@
+	fi
+	cd $(LLVM_REPO)  &&                                                   \
+	mkdir build_libunwind && cd build_libunwind  &&                       \
+	cmake -D musl=Yes                                                     \
```
Running cmake from within a makefile is still… odd. Normally, we show how to run cmake and build on the project’s home page (e.g. [here (for pstore)](https://github.com/SNSystems/pstore#building) or [here (for llvm-project-prepo)](https://github.com/SNSystems/llvm-project-prepo#building-the-compiler or [your work on musl-prepo](https://github.com/MaggieYingYi/musl-prepo) which substibutes `./configure` for cmake).

Is it really necessary to run cmake four times (once for each target)? I would imagine that a single run could produce a build suitable for then building any of the libraries.

MaggieYingYi replied:
> 
> 
> Running cmake from within a makefile is still… odd. Normally, we show how to run cmake and build on the project’s home page (e.g. [here (for pstore)](https://github.com/SNSystems/pstore#building) or [here (for llvm-project-prepo)](https://github.com/SNSystems/llvm-project-prepo#building-the-compiler or [your work on musl-prepo](https://github.com/MaggieYingYi/musl-prepo) which substibutes `./configure` for cmake).
> 

Do you mean we could only provide a Wiki page in the llvm-project-prepo to explain how to build LLVM runtime libraries using musl libc library?

MaggieYingYi replied:
> 
> 
>
> Is it really necessary to run cmake four times (once for each target)? I would imagine that a single run could produce a build suitable for then building any of the libraries.
>

There are several reasons for separating the target:

1) we cannot do cmake in a single run when using the musl libc library.

I tried the following command:

```
cmake -D musl=Yes  -D libcxx=Yes  -D CMAKE_BUILD_TYPE=Release  -D LLVM_PATH=../ -D LLVM_CONFIG_PATH=../build/bin/llvm-config -D LIBUNWIND_ENABLE_SHARED=No -D LIBUNWIND_ENABLE_STATIC=Yes  -D COMPILER_RT_DEFAULT_TARGET_ONLY=ON -D COMPILER_RT_DEFAULT_TARGET_ONLY=ON -D COMPILER_RT_BUILD_BUILTINS=ON -D COMPILER_RT_BUILD_SANITIZERS=OFF -D COMPILER_RT_BUILD_XRAY=OFF -D COMPILER_RT_BUILD_LIBFUZZER=OFF -D COMPILER_RT_BUILD_PROFILE=OFF -D libcxxabi_include=-I../build/lib/clang/11.0.0/include/ -D LIBCXXABI_ENABLE_EXCEPTIONS=OFF -D LIBCXXABI_USE_COMPILER_RT=YES  -D LIBCXXABI_ENABLE_SHARED=No -D LIBCXXABI_ENABLE_STATIC=Yes -D LIBCXX_ENABLE_SHARED=No -D LIBCXX_ENABLE_STATIC=Yes -D LIBCXX_CXX_ABI=libcxxabi -D LIBCXX_CXX_ABI_INCLUDE_PATHS=../libcxxabi/include -D LIBCXX_HAS_MUSL_LIBC=ON -D LLVM_ENABLE_LIBCXX=ON -D CMAKE_TOOLCHAIN_FILE=../llvm/utils/repo/repo.cmake -D CMAKE_INSTALL_PREFIX=~/LLVM -DLLVM_ENABLE_PROJECTS="libunwind;compiler-rt;libcxx;libcxxabi;" ../llvm
```

I got the following cmake failure. 
```
CMake Error at cmake/modules/CheckAtomic.cmake:55 (message):
  Host compiler must support std::atomic!
```

2) compiler-rt contains an implementation of an atomics library. By building targets separately, the issue can be avoid.

3) the libunwind.a is  a dependency of the C++ ABI library, and sometimes is a dependency of other runtimes.

4) the libcxx.a is also dependent on the libcxxabi.a when using the musl libc.

The detailed information could be found here: https://clang.llvm.org/docs/Toolchain.html.

paulhuggett replied:
I think that this probably needs further refinement, but your solution to these issues seems to do the trick for the moment.

paulhuggett replied:
> Do you mean we could only provide a Wiki page in the llvm-project-prepo to explain how to build LLVM runtime libraries using musl libc library?

Yes. We’ll obviously need to hook it up to CI soon as well.

paulhuggett commented on llvm/utils/repo/Makefile:
```diff
@@ -0,0 +1,121 @@
+$(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o: libcompiler-rt.a
+$(LLVM)/lib/linux/clang_rt.crtend-x86_64.o: libcompiler-rt.a	
+
+$(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o.elf: $(LLVM)/lib/linux/clang_rt.crtbegin-x86_64.o
+	repo2obj --repo=$(LLVM_REPO)/build_libcompiler-rt/lib/crt/clang.db $< -o $@
+
+$(LLVM)/lib/linux/clang_rt.crtend-x86_64.o.elf: $(LLVM)/lib/linux/clang_rt.crtend-x86_64.o
+	        repo2obj --repo=$(LLVM_REPO)/build_libcompiler-rt/lib/crt/clang.db $< -o $@
```
Why do we convert crtbegin and crtend to ELF here?

The second rule seems to have extra whitespace at the start.

MaggieYingYi replied:
> Please update the comment to explain this a little more clearly.

Added in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/9400f67b65927921e3f5134453335a67197d8fc9.

MaggieYingYi replied:
>
> The second rule seems to have extra whitespace at the start.
>

Thanks, fixed in the commit https://github.com/SNSystems/llvm-project-prepo/commit/7ef52d812cc1e524350d3241e5bdf90077da4b3e.

MaggieYingYi replied:
> 
> 
> Why do we convert crtbegin and crtend to ELF here?
> 

When building the compiler-rt library, the crtbegin and crtend are not included in the libcompiler-rt.a library. The repo link (using link.py) and ar (using archive.py) involve to use the repo2obj to generate the ELF file. Since the crtbegin and crtend are only compiled, no corresponding crtbegin and crtend ELF files are generated. Therefore, crtbegin and crtend are converted ELF files here.

MaggieYingYi commented:
> 
> 
> > I added a new Makefile to build LLVM libraries
> 
> When you say ”LLVM libraries”, which libraries do you mean? 
> 

LLVM libraries refer to the libunwind.a, libcompiler-rt.a, libc++abi.a and libc++.a.

>
>I’m guessing that this makefile doesn’t build the compiler.
>
You are right. It doesn't build the compiler. I added the prerequisite in the Makefile see below:
``` 
# prerequisite: 1) installed the musl-prepo libc library in the $(MUSL) directory.
#               2) built llvm-project-prepo in the $(LLVM_BUILD) directory. 
````

>
> > using the musl-prepo libc library in commit [6c4f10c](https://github.com/SNSystems/llvm-project-prepo/commit/6c4f10ca7cb881e1330eb14bea461620683715d0) and install it in the docker in commit [2a5b14b](https://github.com/SNSystems/llvm-project-prepo/commit/2a5b14bcf80f6bba91381b6b09507c09f571645f).
> 
> Why is it being added to the docker image? (We don't have the code in there, or the equivalent for the other libraries.)
>

Yes, there is no code. I reverted it https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/4b48f662f7b96ad27ab340c613168f3122a9c4b3.

paulhuggett commented on libcxxabi/src/CMakeLists.txt:
```diff
@@ -16,9 +16,14 @@ set(LIBCXXABI_SOURCES
-  private_typeinfo.cpp
 )
 
+# private_typeinfo.cpp needs to be built by enabling rtti.
```
Do you mean ”private_typeinfo.cpp needs to be built with RTTI enabled”?

MaggieYingYi replied:
Thanks, fixed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/7ef52d812cc1e524350d3241e5bdf90077da4b3e.

paulhuggett commented on libcxxabi/src/CMakeLists.txt:
```diff
@@ -16,9 +16,14 @@ set(LIBCXXABI_SOURCES
 )
 
+# private_typeinfo.cpp needs to be built by enabling rtti.
+# Don't include private_typeinfo.cpp in the source since repo doesn't support rtti.
```
> since repo doesn't support rtti

Doesn’t _yet_ support RTTI.

MaggieYingYi replied:
Fixed in the commit: https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/7ef52d812cc1e524350d3241e5bdf90077da4b3e.

MaggieYingYi commented:
> 
> 
> I think that this is a good approach to the problem of supporting the new C/C++ stdlibs. Can I confirm that the instructions for building LLVM and pstore with the docker image are still good with this revised toolchain file?

Yes, I can confirm that the docker image are still built fine with this revised toolchain file.

paulhuggett commented:
> Yes, I can confirm that the docker image are still built fine with this revised toolchain file.

To be clear: I’m not asking about the docker image itself. I want to check that the instructions for building LLVM and pstore _within_ the docker image are still good with this revised toolchain file.

MaggieYingYi commented on llvm/utils/repo/Makefile:
```diff
@@ -0,0 +1,121 @@
+#
+# Makefile for building LLVM libraries using musl-prepo libc (requires GNU make)
```
> 
> 
> Is there a compelling reason why we have this makefile here rather than in a separate repository (as the standard C library is)?
> 

The Makefile is used to build LLVM runtime libraries. All runtime libraries source code is in llvm-project-prepo repository. Therefore I put the Makefile here. Any suggestion? 

> Talking about “building LLVM libraries” is confusing because all of LLVM is a collection of libraries! Perhaps using the phrase “LLVM _runtime_ libraries” more accurately describes them.

Good point, changed in commit: https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/7ef52d812cc1e524350d3241e5bdf90077da4b3e

paulhuggett replied:
> Any suggested repository?

Personally, I would create a single repository which contains the materials and instructions for building all of the runtime libraries.

MaggieYingYi commented:
> Please update the comment to explain this a little more clearly.

Added in the https://github.com/SNSystems/llvm-project-prepo/pull/145/commits/9400f67b65927921e3f5134453335a67197d8fc9.

MaggieYingYi commented:
> 
> 
> > Yes, I can confirm that the docker image are still built fine with this revised toolchain file.
> 
> To be clear: I’m not asking about the docker image itself. I want to check that the instructions for building LLVM and pstore _within_ the docker container are still good with this revised toolchain file.
>

Yes, I built the llvm and pstore in the docker image using the revised toolchain file. It works fine.

MaggieYingYi commented:
>
>>>Why is it being added to the docker image? (We don't have the code in there, or the equivalent for the other libraries.)
>>>
>>   Yes, there is no code. I reverted it 4b48f66.
>>
> Why is it being added to the docker image?
>

Initially, I want to use the Makefile to build the LLVM runtime libraries in the docker image. Therefore, I installed it into the docker. And then I found I need the llvm-project-prepo source file to build LLVM runtime libraries. Once I checkout the source code, the Makefile is included. Therefore, there is no need to install the file in the docker. I have removed it.
